pgjdbc/src/test/java/org/postgresql/test/jdbc4/PGCopyInputStreamTest.java

```diff
@@ -5,111 +5,112 @@
-    insertSomeData();
-
-    sut = new PGCopyInputStream((PGConnection) conn,
-        "COPY (select i from cpinstreamtest order by i asc) TO STDOUT WITH " + copyParams);
-
-    byte[] buf = new byte[2]; // small enough to read in multiple chunks
-    StringBuilder result = new StringBuilder(100);
-    int chunks = 0;
-    while (sut.read(buf) > 0) {
-      result.append(new String(buf));
-      ++chunks;
+    List<String> chunks = new ArrayList<>();
+    PGConnection pgConn = conn.unwrap(PGConnection.class);
+    try (PGCopyInputStream in = new PGCopyInputStream(pgConn, COPY_IN_ORDER_SQL)) {
+      // Allocate size of exactly one row
+      byte[] buf = new byte[COPY_ROW_SIZE];
+      while (in.read(buf) > 0) {
+        chunks.add(new String(buf, StandardCharsets.UTF_8));
+      }
     }
-
-    assertEquals(4, chunks);
-    assertEquals("0\n1\n2\n3\n", result.toString());
+    assertEquals("Should read one chunk per row", NUM_TEST_ROWS, chunks.size());
+    assertEquals("Entire table should have be read", "0\n1\n2\n3\n", TestUtil.join(chunks, ""));
   }
 
   @Test
-  public void testStreamCanBeClosedAfterReadUp() throws SQLException, IOException {
-    insertSomeData();
-
-    sut = new PGCopyInputStream((PGConnection) conn,
-        "COPY (select i from cpinstreamtest order by i asc) TO STDOUT WITH " + copyParams);
-
-    byte[] buff = new byte[100];
-    while (sut.read(buff) > 0) {
-      // do nothing
-    }
-
-    sut.close();
-  }
-
-  private void silentlyCloseStream(PGCopyInputStream sut) {
-    if (sut != null) {
-      try {
-        if (sut.isActive()) {
-          sut.close();
-        }
-      } catch (IOException e) {
-        // intentionally ignore
+  public void testCopyAPI() throws SQLException, IOException {
+    List<String> chunks = new ArrayList<>();
+    PGConnection pgConn = conn.unwrap(PGConnection.class);
+    try (PGCopyInputStream in = new PGCopyInputStream(pgConn, COPY_IN_ORDER_SQL)) {
+      byte[] buf;
+      while ((buf = in.readFromCopy()) != null) {
+        chunks.add(new String(buf, StandardCharsets.UTF_8));
       }
     }
+    assertEquals("Should read one chunk per row", NUM_TEST_ROWS, chunks.size());
+    assertEquals("Entire table should have be read", "0\n1\n2\n3\n", TestUtil.join(chunks, ""));
   }
 
-  private void insertSomeData() throws SQLException {
-    PreparedStatement pstmt = conn.prepareStatement("insert into cpinstreamtest (i) values (?)");
-    for (int i = 0; i < 4; ++i) {
-      pstmt.setInt(1, i);
-      pstmt.addBatch();
+  @Test
+  public void testMixedAPI() throws SQLException, IOException {
+    List<String> chunks = new ArrayList<>();
+    PGConnection pgConn = conn.unwrap(PGConnection.class);
+    try (PGCopyInputStream in = new PGCopyInputStream(pgConn, COPY_IN_ORDER_SQL)) {
+      // First read using java.io.InputStream API
+      byte[] firstChar = new byte[1];
+      in.read(firstChar);
+      assertArrayEquals("IO API should read first character", "0".getBytes(), firstChar);
```

Comment:
Just wondering. Have you tried `new StrangeInputStream(in,...)` + full read from the stream?

Technically speaking, if you read from that `StrangeInputStream` then it would perform random-sized reads to the downstream inputstream, which might uncover more bugs.

Reply:
No I just wanted to add something that mixed the two APIs. It's not so much an issue of reading in random amounts as it is mixing the java.io API that allows for arbitrary sized byte reads and the CopyOut API that returns entire rows.

Reply:
If you want to mix Java APIs, then you should exercise `read()` and `read(byte[] buf, int off, int siz)` methods as well (which `StrangeInputStream` could probably do automatically)